diff -ru msgpack-0.5.7.orig/src/Makefile.am msgpack-0.5.7/src/Makefile.am
--- msgpack-0.5.7.orig/src/Makefile.am	2011-06-12 15:07:29.000000000 +0900
+++ msgpack-0.5.7/src/Makefile.am	2012-02-27 14:40:02.265682155 +0900
@@ -20,7 +20,7 @@
 
 
 # -version-info CURRENT:REVISION:AGE
-libmsgpack_la_LDFLAGS = -version-info 3:0:0
+libmsgpack_la_LDFLAGS = -version-info 3:0:0 -no-undefined
 
 
 # backward compatibility
@@ -33,7 +33,7 @@
 		vrefbuffer.c \
 		zone.c
 
-libmsgpackc_la_LDFLAGS = -version-info 2:0:0
+libmsgpackc_la_LDFLAGS = -version-info 2:0:0 -no-undefined
 
 
 nobase_include_HEADERS = \
