diff -ru nginx-1.9.5.orig/src/core/ngx_connection.c nginx-1.9.5/src/core/ngx_connection.c
--- nginx-1.9.5.orig/src/core/ngx_connection.c	2015-09-28 09:56:45.660075199 +0900
+++ nginx-1.9.5/src/core/ngx_connection.c	2015-09-28 10:28:48.284449372 +0900
@@ -1175,7 +1175,7 @@
 
         /* THREAD: lock */
 
-        if (c[i].fd != -1 && c[i].idle) {
+        if (c[i].fd != (ngx_socket_t) -1 && c[i].idle) {
             c[i].close = 1;
             c[i].read->handler(c[i].read);
         }
